pkg/webhook/admission/pod/storage_initializer_injector.go

```diff
@@ -330,12 +331,26 @@ func (mi *StorageInitializerInjector) InjectStorageInitializer(pod *v1.Pod) erro
 		}
 	}
 
-	// Allow to override the uid for the case where ISTIO CNI with DNS proxy is enabled
-	// See for more: https://istio.io/latest/docs/setup/additional-setup/cni/#compatibility-with-application-init-containers.
-	if value, ok := pod.GetAnnotations()[constants.IstioSidecarUIDAnnotationKey]; ok {
-		if uid, err := strconv.ParseInt(value, 10, 64); err == nil {
-			initContainer.SecurityContext.RunAsUser = ptr.Int64(uid)
-		}
+	/*
+		OpenShift uses istio-cni which causes an issue with init-containers when calling external services
+		like S3 or similar. Setting the `uid` for the `storage-initializer` to the same `uid` as the
+		`uid` of the `istio-proxy` resolves the issue. In OpenShift the `istio-proxy` always gets assigned
+		the first `uid` from the namespaces `uid` range + 1 (The range is defined in an annotation on the namespace).
+	*/
+	if initContainer.SecurityContext == nil {
+		initContainer.SecurityContext = &v1.SecurityContext{}
+	}
+	uidStr := targetNs.Annotations[OpenShiftUidRangeAnnotationKey]
```

Comment:
Hm not sure if it is better to fail or not (if the annotation is not there, something went pretty wrong on OCP). As the user in OCP probably will never know about the issue and how to resolve it (needing to know to add 1 for example), do we expect that he even could bring the correct user? I'm currently favouring erroring I think.

Reply:
I don't think that this mechanism with "+1" would work as Istio itself would not find that range, right ? I'm not sure what Istio/Service Mesh would do as a fallback when the range is not available as an annotation. I would then probably do the same here, too.

Reply:
Yeah, I think it should be fine to assume it is there or error out.

Reply:
I think I'm more in agreement with @danielezonca here.

I think the code should honor the `IstioSidecarUIDAnnotationKey` if present. Otherwise, apply OpenShift logic. If both the OpenShift annotations and `IstioSidecarUIDAnnotationKey` aren't there, continue without doing any change and without error .

That logic ^ would make the code compatible with both upstream Istio-cni (which will want the `1337` annotation) and Maistra/OSSM (which will use OpenShift annotations). And the code would also remain compatible with an upstream non-cni-Istio installation (which presumably works well without UID hacking), in case community wants to try that.

Reply:
...I'm also thinking that such logic ^ (despite being a workaround) is perhaps more feasible to push to upstream, as it is less invasive.

Reply:
I like your proposal and explanation. I think that makes sense. I updated the PR accordingly, please re-review.